### PR TITLE
Feat: `jenkins` plugin State Management Enhancement

### PR DESCRIPTION
**Pull request by daniel-hutao**
## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description

1. `jenkins` plugin state management enhancement;
2. add `outputs` to `jenkins` plugin;
3. unify the way "outputs" are set up of all plugins;
4. docs updated.

## Related Issues

#993

## New Behavior (screenshots if needed)

- config `config-jenkins.yaml`

```yaml
---
# core config
varFile: "" # If not empty, use the specified external variables config file
toolFile: "" # If not empty, use the specified external tools config file
pluginDir: ".devstream"
state: # state config, backend can be local or s3
  backend: local
  options:
    stateFile: devstream.state

---
tools:
  # name of the tool
  - name: jenkins
    # id of the tool instance
    instanceID: default
    # format: name.instanceID; If specified, dtm will make sure the dependency is applied first before handling this tool.
    dependsOn: [ ]
    # options for the plugin
    options:
      chart:
        # custom configuration. You can refer to [Jenkins values.yaml](https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/values.yaml)
        values_yaml: |
          serviceAccount:
            create: true
            name: jenkins
          controller:
            adminUser: "admin"
            ingress:
              enabled: true
              hostName: jenkins.example.com
```

- apply once `./dtm apply -f config-jenkins.yaml -y`

```sh
2022-08-23 11:17:22 ℹ [INFO]  Apply started.
2022-08-23 11:17:22 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-08-23 11:17:22 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-08-23 11:17:22 ℹ [INFO]  Tool (jenkins/default) found in config but doesn't exist in the state, will be created.
2022-08-23 11:17:22 ℹ [INFO]  Start executing the plan.
2022-08-23 11:17:22 ℹ [INFO]  Changes count: 1.
2022-08-23 11:17:22 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-08-23 11:17:22 ℹ [INFO]  Processing: (jenkins/default) -> Create ...
2022-08-23 11:17:23 ℹ [INFO]  Creating or updating helm chart ...
2022/08/23 11:17:26 creating 13 resource(s)
2022/08/23 11:17:27 beginning wait for 13 resources with timeout of 5m0s
2022/08/23 11:17:27 StatefulSet is not ready: jenkins/jenkins. 0 out of 1 expected pods are ready
2022/08/23 11:17:29 StatefulSet is not ready: jenkins/jenkins. 0 out of 1 expected pods are ready
...
2022/08/23 11:19:57 StatefulSet is not ready: jenkins/jenkins. 0 out of 1 expected pods are ready
2022/08/23 11:19:59 release installed successfully: jenkins/jenkins-4.1.17
2022-08-23 11:19:59 ✔ [SUCCESS]  Tool (jenkins/default) Create done.
2022-08-23 11:19:59 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-08-23 11:19:59 ✔ [SUCCESS]  All plugins applied successfully.
2022-08-23 11:19:59 ✔ [SUCCESS]  Apply finished.
```

- apply twice `./dtm apply -f config-jenkins.yaml -y`

```sh
2022-08-23 11:20:15 ℹ [INFO]  Apply started.
2022-08-23 11:20:15 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-08-23 11:20:16 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-08-23 11:20:16 ℹ [INFO]  No changes done since last apply. There is nothing to do.
2022-08-23 11:20:16 ✔ [SUCCESS]  Apply finished.
```

- state `cat devstream.state`

```yaml
jenkins_default:
  name: jenkins
  instanceid: default
  dependson: []
  options:
    chart:
      values_yaml: |
        serviceAccount:
          create: true
          name: jenkins
        controller:
          adminUser: "admin"
          ingress:
            enabled: true
            hostName: jenkins.example.com
  resource:
    outputs:
      jenkins_url: http://jenkins.jenkins:8080
    values_yaml: |
      serviceAccount:
        create: true
        name: jenkins
      controller:
        adminUser: "admin"
        ingress:
          enabled: true
          hostName: jenkins.example.com
    workflows: |
      statefulsets:
        - name: jenkins
          ready: true
```

- delete `./dtm delete -f config-jenkins.yaml -y`

```sh
2022-08-23 11:21:42 ℹ [INFO]  Delete started.
2022-08-23 11:21:42 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-08-23 11:21:42 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-08-23 11:21:42 ℹ [INFO]  Tool (jenkins/default) will be deleted.
2022-08-23 11:21:42 ℹ [INFO]  Start executing the plan.
2022-08-23 11:21:42 ℹ [INFO]  Changes count: 1.
2022-08-23 11:21:42 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-08-23 11:21:42 ℹ [INFO]  Processing: (jenkins/default) -> Delete ...
2022-08-23 11:21:43 ℹ [INFO]  Uninstalling jenkins helm chart.
2022/08/23 11:21:43 uninstall: Deleting jenkins
2022/08/23 11:21:44 Starting delete for "jenkins" Ingress
2022/08/23 11:21:44 Starting delete for "jenkins-agent" Service
2022/08/23 11:21:44 Starting delete for "jenkins" Service
2022/08/23 11:21:44 Starting delete for "jenkins" StatefulSet
2022/08/23 11:21:44 Starting delete for "jenkins-watch-configmaps" RoleBinding
2022/08/23 11:21:44 Starting delete for "jenkins-schedule-agents" RoleBinding
2022/08/23 11:21:44 Starting delete for "jenkins-casc-reload" Role
2022/08/23 11:21:44 Starting delete for "jenkins-schedule-agents" Role
2022/08/23 11:21:44 Starting delete for "jenkins" PersistentVolumeClaim
2022/08/23 11:21:44 Starting delete for "jenkins-jenkins-jcasc-config" ConfigMap
2022/08/23 11:21:44 Starting delete for "jenkins" ConfigMap
2022/08/23 11:21:44 Starting delete for "jenkins" Secret
2022/08/23 11:21:44 Starting delete for "jenkins" ServiceAccount
2022/08/23 11:21:44 purge requested for jenkins
2022/08/23 11:21:44 release uninstalled, response: &{0x140005edd50 }
2022-08-23 11:21:44 ℹ [INFO]  Prepare to delete 'jenkins_default' from States.
2022-08-23 11:21:44 ✔ [SUCCESS]  Tool (jenkins/default) delete done.
2022-08-23 11:21:44 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-08-23 11:21:44 ✔ [SUCCESS]  All plugins deleted successfully.
2022-08-23 11:21:44 ✔ [SUCCESS]  Delete finished.
```

### CODE REVIEW

File: internal/pkg/plugin/jenkins/jenkins.go [hunk 1]
```diff
@@ -23,24 +23,26 @@ var defaultHelmConfig = helm.Options{
...
+func genJenkinsState(options plugininstaller.RawOptions) (statemanager.ResourceState, error) {
+	resState, err := helm.GetPluginAllState(options)
 	if err != nil {
 		return nil, err
 	}
 
-	outputs := map[string]interface{}{}
+	// svc_name.svc_ns:svc_port
+	outputs := map[string]interface{}{
+		"jenkins_url": "http://jenkins.jenkins:8080",
```
Comment (aFlyBird0) [inline internal/pkg/plugin/jenkins/jenkins.go:34]: note: service name could be changed if the user changes the default configuration.

File: internal/pkg/plugin/jenkins/jenkins.go [hunk 2]
```diff
@@ -23,24 +23,26 @@ var defaultHelmConfig = helm.Options{
...
-	outputs["foo"] = "bar"
+	// values.yaml
+	opt, err := helm.NewOptions(options)
+	if err != nil {
+		return nil, err
+	}
 
-	resource["outputs"] = outputs
+	valuesYaml := opt.GetHelmParam().Chart.ValuesYaml
+	resState["values_yaml"] = valuesYaml
```
Comment (aFlyBird0) [inline internal/pkg/plugin/jenkins/jenkins.go:45]: Maybe you could update the `helm installer` directly?

**Review by aFlyBird0 — APPROVED** [review]

File: internal/pkg/plugin/jenkins/jenkins.go (hunk 1 above)
Comment (daniel-hutao) [reply internal/pkg/plugin/jenkins/jenkins.go:34]: > note: service name could be changed if the user changes the default configuration.

I know it. But users can't change the svc name from values.yaml and it's no meaning.

Comment (aFlyBird0) [reply internal/pkg/plugin/jenkins/jenkins.go:34]: The svc name of Jenkins is determined by chart release name.

image

Comment (aFlyBird0) [reply internal/pkg/plugin/jenkins/jenkins.go:34]:
```go
func (opts *jenkinsOptions) getJenkinsFullName() string {
	if strings.Contains(opts.Chart.ChartName, opts.Chart.ReleaseName) {
		return opts.Chart.ReleaseName
	}

	return fmt.Sprintf("%s-%s", opts.Chart.ReleaseName, opts.Chart.ChartName)
}
```

Comment (daniel-hutao) [reply internal/pkg/plugin/jenkins/jenkins.go:34]: get it!

Comment (daniel-hutao) [reply internal/pkg/plugin/jenkins/jenkins.go:34]: I'll fix it in the next pr. Thanks for the input!

**Pull request merged**